File: solutions/leetcode_931.py
from functools import cache


class Solution:
    def minFallingPathSum(self, matrix: list[list[int]]) -> int:
        @cache
        def helper(col, row):
            if col < 0 or col == len(matrix[0]):
                return 1_000_000
            if row == len(matrix):
                return 0
            return matrix[row][col] + min(
                helper(col - 1, row + 1),
                helper(col, row + 1),
                helper(col + 1, row + 1)
            )
        return min(helper(col, 0)
            for col in range(len(matrix))
        )


if __name__ == '__main__':
    given_matrix = [[-19,57],[-40,-5]]
    expected = -59

    solution = Solution()
    result = solution.minFallingPathSum(given_matrix)
    assert result == expected

File: leetcode_931.c
int minFallingPathSum(int** matrix, int matrixSize, int* matrixColSize)
{
    for (int row = 1; row < matrixSize; row++) {
        for (int col = 0; col < matrixSize; col++) {
            int min = matrix[row - 1][col];
            if (col - 1 >= 0 && matrix[row - 1][col - 1] < min)
                min = matrix[row - 1][col - 1];
            if (col + 1 < matrixSize && matrix[row - 1][col + 1] < min)
                min = matrix[row - 1][col + 1];
            matrix[row][col] += min;
        }
    }
    int min = INT_MAX;
    int row = matrixSize - 1;
    for (int col = 0; col < matrixSize; col++) {
        if (matrix[row][col] < min)
            min = matrix[row][col];
    }

    return min;
}
